Add 'eigmin'/'eigmax' methods for 'Eigen'

Luis-Varona opened the pull request:
We add the methods 'eigmin' and 'eigmax' on the type 'F::Union{Eigen, GeneralizedEigen}' similarly to how 'eigvals' is defined on 'F'. Note that the corresponding unit tests in 'test/eigen.jl' that call 'eigvals' on such a struct are updated to include 'eigmin' and 'eigmax' calls as well; if all eigenvalues are real, the actual functionality of the methods is tested, and if any happen to be complex, then we test that they throw 'MethodError's.

(Resolves #1464 by @longemen3000.)

Luis-Varona committed:
Add 'eigmin'/'eigmax' methods for 'Eigen'

We add the methods 'eigmin' and 'eigmax' on the type 'F::Union{Eigen, GeneralizedEigen}' similarly to how 'eigvals' is defined on 'F'. Note that the corresponding unit tests in 'test/eigen.jl' that call 'eigvals' on such a struct are updated to include 'eigmin' and 'eigmax' calls as well; if all eigenvalues are real, the actual functionality of the methods is tested, and if any happen to be complex, then we test that they throw 'MethodError's.

(Resolves #1464 by @longemen3000.)

jishnub commented on src/eigen.jl:
```diff
@@ -276,6 +276,8 @@ eigvecs(A::Union{Number, AbstractMatrix}; kws...) =
 eigvecs(F::Union{Eigen, GeneralizedEigen}) = F.vectors
 
 eigvals(F::Union{Eigen, GeneralizedEigen}) = F.values
+eigmin(F::Union{Eigen, GeneralizedEigen}) = minimum(F.values)
```
```suggestion
eigmin(F::Union{Eigen, GeneralizedEigen}) = minimum(eigvals(F))
```

jishnub commented on src/eigen.jl:
```diff
@@ -276,6 +276,8 @@ eigvecs(A::Union{Number, AbstractMatrix}; kws...) =
 
 eigvals(F::Union{Eigen, GeneralizedEigen}) = F.values
+eigmin(F::Union{Eigen, GeneralizedEigen}) = minimum(F.values)
+eigmax(F::Union{Eigen, GeneralizedEigen}) = maximum(F.values)
```
```suggestion
eigmax(F::Union{Eigen, GeneralizedEigen}) = maximum(eigvals(F))
```

jishnub commented on test/eigen.jl:
```diff
@@ -44,6 +44,15 @@ aimg  = randn(n,n)/2
             @test eigvals(f) === f.values
+
+            if all(isreal, f.values)
+                @test eigmin(f) == minimum(f.values)
```
```suggestion
                @test eigmin(f) == minimum(eigvals(f))
```

jishnub commented on test/eigen.jl:
```diff
@@ -44,6 +44,15 @@ aimg  = randn(n,n)/2
+
+            if all(isreal, f.values)
+                @test eigmin(f) == minimum(f.values)
+                @test eigmax(f) == maximum(f.values)
```
```suggestion
                @test eigmax(f) == maximum(eigvals(f))
```

jishnub commented on test/eigen.jl:
```diff
@@ -81,6 +90,15 @@ aimg  = randn(n,n)/2
             @test eigvals(f) === f.values
+
+            if all(isreal, f.values)
+                @test eigmin(f) == minimum(f.values)
```
```suggestion
                @test eigmin(f) == minimum(eigvals(f))
```

jishnub commented on test/eigen.jl:
```diff
@@ -81,6 +90,15 @@ aimg  = randn(n,n)/2
+
+            if all(isreal, f.values)
+                @test eigmin(f) == minimum(f.values)
+                @test eigmax(f) == maximum(f.values)
```
```suggestion
                @test eigmax(f) == maximum(eigvals(f))
```

jishnub commented on test/eigen.jl:
```diff
@@ -44,6 +44,15 @@ aimg  = randn(n,n)/2
             @test isposdef(a) == isposdef(f)
             @test eigvals(f) === f.values
+
+            if all(isreal, f.values)
```
```suggestion
            if all(isreal, eigvals(f))
```

jishnub commented on test/eigen.jl:
```diff
@@ -81,6 +90,15 @@ aimg  = randn(n,n)/2
             @test eigvecs(asym_sg, ASG2) == f.vectors
             @test eigvals(f) === f.values
+
+            if all(isreal, f.values)
```
```suggestion
            if all(isreal, eigvals(f))
```

Luis-Varona committed:
Change 'F.values' calls to 'eigvals(F)'

This commit incorporates changes suggested by @jishnub to edit all associated 'F.values'/'f.values' calls in the diff to 'eigvals(F)'/'eigvals(f)' instead.

Co-authored-by: Jishnu Bhattacharya <jishnub.github@gmail.com>

Luis-Varona commented:
@jishnub, I've made the requested changes. Thanks for the review.

Luis-Varona commented:
Hi @jishnub - do you think this is all ready to be merged now? :)

jishnub commented:
Thanks!

hz-xiaxz commented:
Should these functions be doced?

Luis-Varona commented:
> Should these functions be doced?

I was wondering that too, but the already existing `eigvals` method on `F:::Union{Eigen, GeneralizedEigen}` that inspired this in the first place is not doc'ed either. Either all three should be or none should be. Thoughts, @jishnub?

Luis-Varona commented:
@hz-xiaxz, @jishnub - I think I will work on a PR later today that adds these methods for `eigvals`/`eigmax`/`eigmin` to their respective function docstrings.
